##K-Nearest Neighbour on Titanic Dataset

### Importing the required libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

###Uploading the titanic dataset

In [2]:
from google.colab import files
uploaded = files.upload()


Saving gender_submission.csv to gender_submission.csv
Saving test.csv to test.csv
Saving train.csv to train.csv


In [105]:
train_data = pd.read_csv("train.csv")
x_test = pd.read_csv("test.csv")
y_test = pd.read_csv("gender_submission.csv")

In [106]:
x_train = train_data.drop(["Survived"], axis = 1)
y_train = train_data["Survived"]
y_test = y_test.drop(["PassengerId"], axis =1)

In [107]:
y_test=np.array(y_test).reshape(418)

In [108]:
x_train

PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1    female  38.0      1      0          PC 17599  71.2833   C85        C  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3    female  35.0      1      0            113803  53.1000  C123        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

### Performing analysis of attributes.

In [109]:
x_train["Pclass"].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [110]:
x_train["Parch"].value_counts()

0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: Parch, dtype: int64

In [111]:
x_train["Ticket"].value_counts()

347082      7
CA. 2343    7
1601        7
3101295     6
CA 2144     6
           ..
9234        1
19988       1
2693        1
PC 17612    1
370376      1
Name: Ticket, Length: 681, dtype: int64

In [112]:
x_train["Cabin"].value_counts()

B96 B98        4
G6             4
C23 C25 C27    4
C22 C26        3
F33            3
              ..
E34            1
C7             1
C54            1
E36            1
C148           1
Name: Cabin, Length: 147, dtype: int64

In [113]:
x_train["Embarked"].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

### Dropping variables PassengerID, Name, Ticket, Cabin (as ticket and cabin are random names which are neither discrete nor numerical values).

In [114]:
x_train = x_train.drop(["PassengerId", "Name", "Ticket", "Cabin"], axis = 1)
x_test = x_test.drop(["PassengerId", "Name", "Ticket", "Cabin"], axis = 1)

In [115]:
x_train

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         3    male  22.0      1      0   7.2500        S
1         1  female  38.0      1      0  71.2833        C
2         3  female  26.0      0      0   7.9250        S
3         1  female  35.0      1      0  53.1000        S
4         3    male  35.0      0      0   8.0500        S
..      ...     ...   ...    ...    ...      ...      ...
886       2    male  27.0      0      0  13.0000        S
887       1  female  19.0      0      0  30.0000        S
888       3  female   NaN      1      2  23.4500        S
889       1    male  26.0      0      0  30.0000        C
890       3    male  32.0      0      0   7.7500        Q

[891 rows x 7 columns]

In [116]:
x_test

Pclass     Sex   Age  SibSp  Parch      Fare Embarked
0         3    male  34.5      0      0    7.8292        Q
1         3  female  47.0      1      0    7.0000        S
2         2    male  62.0      0      0    9.6875        Q
3         3    male  27.0      0      0    8.6625        S
4         3  female  22.0      1      1   12.2875        S
..      ...     ...   ...    ...    ...       ...      ...
413       3    male   NaN      0      0    8.0500        S
414       1  female  39.0      0      0  108.9000        C
415       3    male  38.5      0      0    7.2500        S
416       3    male   NaN      0      0    8.0500        S
417       3    male   NaN      1      1   22.3583        C

[418 rows x 7 columns]

### Converting Sex, Embarked into numerical values

In [117]:
x_train['Sex'].replace(to_replace=['male', 'female'], value=[1, 0], inplace=True)
x_test['Sex'].replace(to_replace=['male', 'female'], value=[1, 0], inplace=True)
x_train['Embarked'].replace(to_replace=['S', 'C', 'Q'], value=[1, 2, 3], inplace=True)
x_test['Embarked'].replace(to_replace=['S', 'C', 'Q'], value=[1, 2, 3], inplace=True)

In [118]:
x_train["Age"].isna().sum()


177

In [119]:
x_train["Fare"].isna().sum()

0

### There are 177 NaN values in Age variable hence, replacing them with the mean value

In [120]:
x_train["Age"] = x_train["Age"].fillna(x_train["Age"].mean())

In [121]:
x_train

Pclass  Sex        Age  SibSp  Parch     Fare  Embarked
0         3    1  22.000000      1      0   7.2500       1.0
1         1    0  38.000000      1      0  71.2833       2.0
2         3    0  26.000000      0      0   7.9250       1.0
3         1    0  35.000000      1      0  53.1000       1.0
4         3    1  35.000000      0      0   8.0500       1.0
..      ...  ...        ...    ...    ...      ...       ...
886       2    1  27.000000      0      0  13.0000       1.0
887       1    0  19.000000      0      0  30.0000       1.0
888       3    0  29.699118      1      2  23.4500       1.0
889       1    1  26.000000      0      0  30.0000       2.0
890       3    1  32.000000      0      0   7.7500       3.0

[891 rows x 7 columns]

In [122]:
sample1 = x_train.iloc[0]
sample1  = np.array(sample1)
sample1

array([ 3.  ,  1.  , 22.  ,  1.  ,  0.  ,  7.25,  1.  ])

In [123]:
sample2 = x_train.iloc[1]
sample2  = np.array(sample2)
sample2

array([ 1.    ,  0.    , 38.    ,  1.    ,  0.    , 71.2833,  2.    ])

In [124]:
np.sqrt(np.sum(np.square(sample1-sample2)))

66.0474337797465

In [125]:
y_train

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

### Using euclidean distance measure for calculating distance between two data samples

In [126]:
def euclidean_distance(row1, row2):
  row1 = np.array(row1)
  row2 = np.array(row2)
  distance = np.sum(np.square(row1-row2))
  distance = np.sqrt(distance)
  return distance

### Creating an auxiliary function to calculate and return the distance for each data sample

In [127]:
def single_prediction(x_train, y_train, test_sample, num_neighbour):
  x_train = np.array(x_train)
  y_train = np.array(y_train)
  test_sample = np.array(test_sample)
  distances = list()
  for i in range(len(x_train)):
    distances.append(((euclidean_distance(test_sample, x_train[i])),y_train[i]))
    distances.sort(key = lambda tup: tup[0])
  neighbors = list()
  for i in range(num_neighbour):
    neighbors.append(distances[i][1])
  prediction = max(set(neighbors), key = neighbors.count)
  return prediction

In [128]:
results = single_prediction(x_train, y_train, x_test.iloc[0], 4 )

In [129]:
results

0

### Function for predicting labels

In [130]:
def knn_algorithm(x_train, y_train, x_test, num_neighbour):
  x_test = np.array(x_test)
  predictions = list()
  for i in range(len(x_test)):
    test_sample = x_test[i]
    result = single_prediction(x_train, y_train, test_sample, num_neighbour)
    predictions.append(result)
  return predictions

### Function to calculate the accuracy metric of the predicted labels

In [133]:
def accuracy_metric(predictions, actual):
  abs_sum  = np.sum(np.abs(np.array(predictions) - np.array(actual)))
  accuracy = 1-(abs_sum/len(actual))
  return accuracy

### Analysing the KNN algorithm with number of neighbours set to 4

In [131]:
final_prediction = knn_algorithm(x_train, y_train, x_test, 4)

In [132]:
final_prediction = np.array(final_prediction)

In [134]:
accuracy = accuracy_metric(final_prediction, y_test)

In [135]:
accuracy

0.5430622009569378

### Analyzing the KNN algorithm with number of neighbors set to 6

In [139]:
final_prediction_1 = knn_algorithm(x_train, y_train, x_test, 6)

In [140]:
final_prediction_1 = np.array(final_prediction_1)
accuracy = accuracy_metric(final_prediction_1, y_test)

In [141]:
accuracy

0.6507177033492824

### Analyzing the KNN algorithm with number of neighbors set to 8

In [142]:
final_prediction_2 = knn_algorithm(x_train, y_train, x_test, 8)

In [143]:
final_prediction_2 = np.array(final_prediction_2)
accuracy = accuracy_metric(final_prediction_2, y_test)

In [144]:
accuracy

0.65311004784689

### Creating a loop and printing the accuract obtained from setting different number of neigbhours

In [146]:
for i in range (1, 21):
  predicted_labels = knn_algorithm(x_train, y_train, x_test, i)
  predicted_labels = np.array(predicted_labels)
  accuracy = accuracy_metric(predicted_labels, y_test)
  print("Accuracy for" + str(i) + " neighbours is: "+ str(accuracy) )

Accuracy for1 neighbours is: 0.6100478468899522
Accuracy for2 neighbours is: 0.6578947368421053
Accuracy for3 neighbours is: 0.5334928229665072
Accuracy for4 neighbours is: 0.5430622009569378
Accuracy for5 neighbours is: 0.5526315789473684
Accuracy for6 neighbours is: 0.6507177033492824
Accuracy for7 neighbours is: 0.6483253588516746
Accuracy for8 neighbours is: 0.65311004784689
Accuracy for9 neighbours is: 0.6507177033492824
Accuracy for10 neighbours is: 0.6698564593301435
Accuracy for11 neighbours is: 0.5717703349282297
Accuracy for12 neighbours is: 0.5598086124401913
Accuracy for13 neighbours is: 0.5406698564593302
Accuracy for14 neighbours is: 0.6363636363636364
Accuracy for15 neighbours is: 0.6411483253588517
Accuracy for16 neighbours is: 0.6459330143540669
Accuracy for17 neighbours is: 0.6363636363636364
Accuracy for18 neighbours is: 0.6578947368421053
Accuracy for19 neighbours is: 0.6555023923444976
Accuracy for20 neighbours is: 0.6602870813397129


### Hence, the best accuracy we get when we set the number of neighbours to 10